src/secrets.rs

```diff
@@ -1,22 +1,143 @@
 //! Types for working with registry access secrets
 
+use serde::Deserialize;
+
+use base64;
+use dirs;
+use std::collections::HashMap;
+use std::env;
+use std::fs::File;
+use std::io::BufReader;
+use std::path::Path;
+use std::process::{Command, Stdio};
+use std::str;
+use tracing::debug;
+
 /// A method for authenticating to a registry
 pub enum RegistryAuth {
     /// Access the registry anonymously
     Anonymous,
+
     /// Access the registry using HTTP Basic authentication
     Basic(String, String),
+
+    /// Access the registry using credentials configured in the Docker config file
+    Default,
 }
 
 pub(crate) trait Authenticable {
-    fn apply_authentication(self, auth: &RegistryAuth) -> Self;
+    fn apply_authentication(self, registry: &str, auth: &RegistryAuth) -> Self;
 }
 
 impl Authenticable for reqwest::RequestBuilder {
-    fn apply_authentication(self, auth: &RegistryAuth) -> Self {
+    fn apply_authentication(self, registry: &str, auth: &RegistryAuth) -> Self {
         match auth {
             RegistryAuth::Anonymous => self,
             RegistryAuth::Basic(username, password) => self.basic_auth(username, Some(password)),
+            RegistryAuth::Default => match docker_auth(registry) {
+                Ok(userpass) => self.basic_auth(userpass.0, Some(userpass.1)),
+                Err(e) => {
+                    // This effectively falls back to anonymous, instead of raising the error.
+                    // TODO: Raise this error so clients can *require* Docker config auth.
+                    debug!("failed to get credentials: {}", e);
+                    self
+                }
+            },
+        }
+    }
+}
+
+fn docker_auth(registry: &str) -> Result<(String, String), anyhow::Error> {
+    let cfg = get_docker_config()?;
+
+    // TODO: Check any per-repo configured auth, instead of just per-registry.
+
+    let mut acr = cfg.auths.get(registry);
+    // For historical purposes, this key is also accepted for DockerHub.
+    if acr.is_none() && registry == "index.docker.io" {
+        acr = cfg.auths.get("https://index.docker.io/v1/")
+    }
+    if !acr.is_none() {
+        let ac = acr.unwrap();
+
+        if ac.auth != "" {
+            // base64-decode the auth to get username and password.
+            let dec = &base64::decode(&ac.auth)?;
+            let vec = str::from_utf8(dec)?.split(":").collect::<Vec<_>>();
```

Comment:
There is a `split_once` function that you can use here so you don't have to check the length